Avoid obstacles and control the car

Spidergirl93 opened the pull request:
Car avoids obstacles by changing direction and is fully controlled by the user

Spidergirl93 committed:
Update smartcar.ino

Car avoids obstacles by changing direction and is fully controlled by the user

platisd commented on arduino/smartcar/smartcar.ino:
```diff
@@ -1,9 +1,65 @@
+     car.setAngle(cAngle);
+     delay(600);                                                      //This delay is needed for the car to turn in a short while and then go back to its straight direction,  
+   }                                                                  // because we dont want the car to to turn around itself for no reason!
+   if (front.getDistance() > 1 && front.getDistance() < 200)          //This is the obstacle treshhold. The car must detect obstacle closer than 200
```
Each call to getDistance is a NEW measurement. This means that if you call it twice, there may very well be different distances being returned. Can you see how that will be the cause of problems?
How would you fix it?

Spidergirl93 replied:
I stored the distance.getFront() in a variable and used that variable in the rest of the code to avoid the problem you mentioned. Although I think the previous version was more accurate since it was measuring the distance between car and the obstacle in each instant and since the car is moving, every moment is valuable. Of course I haven't thought about it before you mention the problem, but am I thinking correctly?

platisd replied:
Yes, your thought is correct in regards to using a single variable for both.
I wouldn't say it's more accurate. If you want to measure if a number is between a particular range (larger than 0 smaller than something) then you should be using the same number, not two different ones 👍

Spidergirl93 committed:
Obstacle avoiding car

The value of front.getDistance is saved in another variable to avoid different values in each moment.

platisd commented on arduino/smartcar/smartcar.ino:
```diff
@@ -1,9 +1,66 @@
+     car.setAngle(cAngle);
+     delay(600);                                                      //This delay is needed for the car to turn in a short while and then go back to its straight direction,  
+   }                                                                  // because we dont want the car to to turn around itself for no reason!
+   if (distance > 1 && distance < 200)          //This is the obstacle treshhold. The car must detect obstacle closer than 200
+   {                                                                  // and avoid them
+     handleObstacle();
+   }  
```
This doesn't work as expected if the sensors return `0` which is their way of saying "no obstacle detected".

platisd reviewed (commented):
Aside of the problem that I mentioned in the logic, there's **a lot** of duplicated code which will make your efforts harder down the road.
I advise you to refactor this ASAP.
